### DISCUSSION TITLE
PiSound as DAC?

### DISCUSSION THREAD

**Discussion by lazzarello** [probonopd/MiniDexed#149] [Q&A]
This project came up in [a topic on the PiSound forum](https://community.blokas.io/t/bare-metal-pisound/2966/7?u=lazzarello) regarding working with a DAC outside of a Linux kernel. I installed the latest SD card release but the DAC on the PiSound doesn't work by default. Could be interesting to add support for it's audio out + 5 pin DIN MIDI connectors.

**Comment by probonopd** [comment]
Hello @lazzarello.

Currently PCM5102A and PCM5122 based DACs are supported, and WM8960 is in the works.

From looking at https://blokas.io/pisound/ I can't even figure out which chipset it uses. It is not even mentioned on https://blokas.io/pisound/docs/specs/. Do you know?

**Reply by plcktzh** [reply]
According to https://blokas.io/pisound/docs/audio/ it uses the PCM5102A

Edit: Missed your comment below. Nvm.

**Comment by craigyjp** [comment]
I heard from a guy on Facebook that is building 8 of these onto Zero devices that the UDA1334 24/96 DAC works perfectly well.

**Comment by probonopd** [comment]
Acording to https://github.com/dwhinham/mt32-pi/wiki/I²S-DACs the UDA1334 might be 1:1 compatible with the PCM5xxx and might hence work on MiniDexed as is.

According to the same source, the  Blokas Pisound uses a PCM5102A so it _should_ work, too. But I don't have one to test.

**Reply by CircuV** [reply]
Tried putting a working SD Card with minidexed (on a Raspi 2 B ) in my pisound Raspi (Raspi4) and this is not working. Maybe to get it to work with pisound interface a different config ist needed.  It would be nice to see pispound working (with all the additional MIDI IO) it offers.

**Reply by luisfcorreia** [reply]
have you adjusted configuration from PWM to I2S in minidexed.ini?

**Reply by CircuV** [reply]
Yep :)  (From HDMI to I2S in my case)

**Reply by plcktzh** [reply]
As far as I can tell, the pinout used by MiniDexed is incompatible with the one used by Pisound, so it probably will not work out of the box. I'm not sure whether this would work even with the options in minidexed.ini regarding the GPIO pins for the LCD and the encoder, as there seem to be too few free pins left by Pisound. I've asked Blokas about this, still waiting for them to get back.
Meanwhile, I've compiled both pinouts in a spreadsheet to the best of my knowledge - which isn't that much ... Feel free to add a comment and I'll correct and/or add the data there - https://docs.google.com/spreadsheets/d/13idRy4GvI-_3qN9Z5o31UtmQp2KlS9HrBHYXExvaZ6w/edit?usp=sharing

**Reply by lazzarello** [reply]
Does this mean that the pisound GPIO pins are physically unable to be used for any other function regardless of software pin configuration? ~~The version of the board I got has 14 holes exposed on the top of the pisound board, which (I think) I soldered headers into. Checking out if there's any documentation as to what they expose. I can visually see a number of traces that are going to what could be some of the GPIO pins.~~

[There are 7 free pins available](https://blokas.io/pisound/docs/specs/) and the MiniDexed LCD requires 7...so, it could be possible to remove the rotary encoder feature (MIDI program change only, I guess?) and shuffle some pins around to get fancy DAC output.

This is slightly out of my league in that I don't regularly work on embedded stuff but I'm interested enough to try and get a custom image build running locally. Or would this just be changing some text file config?

**Reply by probonopd** [reply]
What can the Pisound do for MiniDexed that a 5 USD DAC can't? Just curious because I didn't spend time looking into "expensive" (> 5 USD) periperhals.

**Reply by plcktzh** [reply]
Its main advantage is its neat all-in-one design, I guess. It has audio and MIDI in/out on board as well as volume and gain pots - the latter of which would probably not be of much use for MiniDexed. Oh, and "The Button", which could probably be used for switching several pre-defined performance configurations in a live situation.
Admittedly, all this could be done with cheaper solutions and some DIY, or even a custom MiniDexed HAT.
For me, I'd mainly like it to work since I've already got one lying around and don't really use it with MODEP or any of the other Patchbox OS modules, since they feel like too much of a hassle to me. Oh, and the audio quality is exceptional.

I've already got a cheap compatible DAC HAT and the other components on the way though. If they should work, I could lend out my Pisound to any developer (preferably in Germany) willing to take on the task of providing Pisound compatibility.

**Reply by lazzarello** [reply]
Three reasons for me

1. I have a PiSound right now.
2. [The audio system has fancy graphs to demonstrate its performance.](https://blokas.io/pisound/docs/audio/)
3. I'm looking to learn more about embedded programming rather then electronics assembly.

I spent some time last night reading about the device tree overlay system. It seems that would be the first place to start? I'm in the western United States so opposite time zone from Germany. Gonna try some hacking today.

**Reply by probonopd** [reply]
I think that the device overlay system does not apply to MiniDexed because we are not using Linux.

**Reply by lazzarello** [reply]
well then, not really sure where to start other then getting the build script to work locally.

**Reply by probonopd** [reply]
Find out the pinout: Of the PCM5102A and the rest of its built-in devices. Which Raspberry Pi pins are they using? then edit `minidexed.ini` accordingly. If that is sufficient, then no source code changes should be needed.

**Reply by muziker** [reply]
Right, so does the pisound work with minidexed?

**Reply by diyelectromusic** [reply]
I'm struggling to find any detailed information about the GPIO usage in pisound (why isn't there a listing in the documentation?  Seems like the most likely thing people will want to know!), so I can't see if it is using the standard I2S pins for the DAC or the "alternative" pins (see the "PCM" pins here: https://elinux.org/RPi_BCM2835_GPIOs).

Also, judging by the fact that MT32-Pi has its own pisound driver, and looking through the code, it would appear that MIDI is accessed over SPI rather than the Pi's UART, and it looks like the sample rate also needs setting using the three "OSR" GPIO pins.

So, in short, I think pisound is almost certainly a "no" at present and would need some work to work with MiniDexed, but without a schematic or a proper listing of GPIO pins it is hard to tell and would need reading of the MT32-Pi source, the linux kernel driver and possibly some reverse engineering of the dts overlay too...

Kevin

**Reply by perhak** [reply]
There is a list in https://blokas.io/pisound/docs/specs/ of the pins used.

**Reply by probonopd** [reply]
Given that capable PCM5102A based DACs can be had around ~5 USD from Aliexpress, maybe using that is the easier (and more budget friendly) option.

**Reply by diyelectromusic** [reply]
> There is a list in https://blokas.io/pisound/docs/specs/ of the pins used.

Yes, I found that of course, but it colours in the pins used, so you know which are free, but unless I'm missing something really obvious, it doesn't actually tell you what any of them are used /for/...?

The later list of pins is for the 14-way pi-sound header.

Kevin
